File: old_python/python/magnetometry/sparse_OU_library/make_operators.py
import numpy as np
from scipy import sparse

def make_operators(size, g, kappa, delta_r, delta_s, kappa_1, beta, gamma_dec, gamma_phi, LOPhi, W_up, W_down):
    sigmam = np.array([[0, 0], [1, 0]])
    sigmaz = np.array([[1, 0], [0, -1]])
    gamma_p = np.zeros(size)
    alpha = np.zeros(size, dtype=complex)
    a_ad = np.zeros((size, 2, 2), dtype=complex)
    eps_s = np.zeros(size)
    H_ap = np.zeros((size, 2, 2))
    for i in range(size):
        gamma_p[i] = 2 * g[i]**2 * kappa[i] / (kappa[i]**2 + (delta_r[i] - delta_s[i])**2)
        alpha[i] = (2 * kappa_1[i])**0.5 * beta[i] / (kappa[i] + 1j * delta_r[i])
        a_ad[i] = alpha[i] * np.identity(2) - 1j * g[i] * sigmam / (kappa[i] + 1j * (delta_r[i] - delta_s[i]))
        eps_s[i] = (delta_r[i] - delta_s[i]) * g[i]**2 * kappa[i] / (kappa[i]**2 + (delta_r[i] - delta_s[i])**2)
        H_ap[i] = (delta_s[i] * sigmaz / 2 + g[i] * (alpha[i] * sigmam.T + alpha[i].conj() * sigmam) - eps_s[i] * sigmam.T @ sigmam).real

    #Block diagonal Hamiltonian and operators
    c_1 = np.zeros((2 * size, 2 * size))
    c_2 = np.zeros((2 * size, 2 * size))
    c_3 = np.zeros((2 * size, 2 * size))
    c = np.zeros((2 * size, 2 * size), dtype=complex)
    H = np.zeros((2 * size, 2 * size))
    Jup = np.zeros((2 * size, 2 * size))
    Jdown = np.zeros((2 * size, 2 * size))
    for i in range(size):
        subrho = np.zeros((size,size))
        subrho[size - i - 1, size - i - 1] = 1
        H += np.kron(subrho, H_ap[i])
        c_1 += np.kron(subrho, gamma_p[i]**0.5 * sigmam)
        c_2 +=  np.kron(subrho, gamma_dec[i]**0.5 * sigmam)
        c_3 += np.kron(subrho, (gamma_phi[i] / 2)**0.5 * sigmaz)
        c += np.kron(subrho, (2 * kappa_1[i])**0.5 * (a_ad[i] - beta[i] * np.identity(2)) * np.exp(-1j * LOPhi))
        if i != size - 1:
            up_element = np.zeros((size, size))
            down_element = np.zeros((size, size))
            up_element[i, i + 1] = 1
            down_element[i + 1, i] = 1
            Jup += np.kron(up_element, W_up[i]**0.5 * np.identity(2))
            Jdown += np.kron(down_element, W_down[i]**0.5 * np.identity(2))
    H = sparse.csr_matrix(H)
    c = sparse.csr_matrix(c)
    c_1 = sparse.csr_matrix(c_1)
    c_2 = sparse.csr_matrix(c_2)
    c_3 = sparse.csr_matrix(c_3)
    Jup = sparse.csr_matrix(Jup)
    Jdown = sparse.csr_matrix(Jdown)
    return c_1, c_2, c_3, c, H, Jup, Jdown

File: make_operators.cpp
#include <vector>
#include <tuple>
#include <cmath>
#include <armadillo>
#include <complex>

using namespace std::complex_literals;

std::tuple< arma::sp_cx_mat, arma::sp_cx_mat, arma::sp_cx_mat, arma::sp_cx_mat, arma::sp_cx_mat, arma::sp_cx_mat, arma::sp_cx_mat > make_operators(const double N, const double g, const double kappa, const double kappa_1, const double delta_c, const std::vector<double> delta_q, const double beta, const double gamma_dec, const double gamma_phi, const double LOPhi, std::vector<double> r_up, std::vector<double> r_down){
        // Initial bullshit, making necessary parameters from initial parameters
        arma::sp_cx_mat sigma_m(2, 2);
	sigma_m(1, 0) = 1.;//	= { {0. + 0i, 0. + 0i}, {1. + 0i, 0. + 0i} };
        arma::sp_cx_mat sigma_z(2, 2);
     	sigma_z(0, 0) = 1.;
	sigma_z(1, 1) = -1.;//	= { {1. + 0i, 0. + 0i}, {0. + 0i, -1. + 0i} };

        std::vector<double> gamma_p;
	for (unsigned int i = 0; i<N; ++i) gamma_p.push_back(2 * g * g * kappa / (kappa * kappa + pow(delta_c - delta_q.at(i), 2)));

        const double alpha = std::real(sqrt(2.0 * kappa_1) * beta / (kappa + 1i * delta_c));

        std::vector< arma::sp_cx_mat > a_ad(N);
        for (int i = 0; i < N; ++i)
               a_ad[i] = alpha * arma::speye(2, 2) - 1i * g * sigma_m / (kappa + 1i * (delta_c - delta_q[i]));
	
        std::vector < double > eps_q(N);
        for (int i = 0; i<N; ++i)
                eps_q[i] = (delta_c - delta_q[i]) * g * g * kappa / ( kappa * kappa + pow(delta_c - delta_q[i] , 2));
	

        std::vector < arma::sp_cx_mat > H_ap;
        for(int i = 0; i<N; ++i)
                H_ap.push_back(delta_q[i] * sigma_z / 2 + g * (alpha * sigma_m.t() + alpha * sigma_m) - eps_q[i] * sigma_m.t() * sigma_m);
	std::reverse(H_ap.begin(), H_ap.end()); //Building H from ground up, so lowest right diagonal is lowest state


        // Making the operators

        // c operators
        arma::sp_cx_mat c_1(2 * N, 2 * N);
        for (int i = 0; i<N; ++i) c_1.submat(2 * i, 2 * i, 2 * i + 1, 2 * i + 1) = sqrt(gamma_p[i]) * sigma_m;
	arma::sp_cx_mat complex_N_N_identity(N, N);
	complex_N_N_identity.speye();
        arma::sp_cx_mat c_2 = arma::kron(complex_N_N_identity, sqrt(gamma_dec) * sigma_m);

        arma::sp_cx_mat c_3 = arma::kron(complex_N_N_identity, sqrt(gamma_phi / 2) * sigma_z);

        arma::sp_cx_mat c(2 * N, 2 * N);
        for (int i = 0; i<N; ++i)
                c.submat(2 * i, 2 * i, 2 * i + 1, 2 * i + 1) = sqrt(2 * kappa_1) * (a_ad.at(i) - beta * arma::speye(2, 2)) * exp(-1i * LOPhi);
	c.clean(1e2 * arma::datum::eps);

        // H
        arma::sp_cx_mat H(2 * N, 2 * N);
        for (int i = 0; i<N; ++i)
                H.submat(2 * i, 2 * i, 2 * i + 1, 2 * i + 1) = H_ap.at(i);


        //Jump up and down operators
        arma::sp_cx_mat J_up(2 * N, 2 * N), J_down(2 * N, 2 * N);
        for (int i = 0; i < (N - 1); ++i){
                J_up.submat(2 * i, 2 * (i + 1), 2 * i + 1, 2 * (i + 1) + 1).eye();
                J_up.submat(2 * i, 2 * (i + 1), 2 * i + 1, 2 * (i + 1) + 1) *= sqrt(r_up.at(i)) ;
                J_down.submat(2 * (i + 1), 2 * i, 2 * (i + 1) + 1, 2 * i + 1).eye();
                J_down.submat(2 * (i + 1), 2 * i, 2 * (i + 1) + 1, 2 * i + 1) *= sqrt(r_down.at(i));
        }
	
        return {c_1, c_2, c_3, c, H, J_up, J_down};
}
